src/Course/FileIO.hs

```diff
@@ -81,54 +81,52 @@ the contents of c
 
 -- Given the file name, and file contents, print them.
 -- Use @putStrLn@.
-printFile ::
-  FilePath
-  -> Chars
-  -> IO ()
-printFile =
-  error "todo: Course.FileIO#printFile"
+printFile :: FilePath -> Chars -> IO ()
+printFile path contents =
+    putStrLn $ "============ " ++ path ++ "\n" ++ contents
 
 -- Given a list of (file name and file contents), print each.
 -- Use @printFile@.
-printFiles ::
-  List (FilePath, Chars)
-  -> IO ()
-printFiles =
-  error "todo: Course.FileIO#printFiles"
+printFiles :: List (FilePath, Chars) -> IO ()
+printFiles files = execIoNoResult printTuple files
+  where
+    printTuple (path, chars) = printFile path chars
 
 -- Given a file name, return (file name and file contents).
 -- Use @readFile@.
-getFile ::
-  FilePath
-  -> IO (FilePath, Chars)
-getFile =
-  error "todo: Course.FileIO#getFile"
+getFile :: FilePath -> IO (FilePath, Chars)
+getFile path = ((,) path) <$> readFile path
 
 -- Given a list of file names, return list of (file name and file contents).
 -- Use @getFile@.
-getFiles ::
-  List FilePath
-  -> IO (List (FilePath, Chars))
-getFiles =
-  error "todo: Course.FileIO#getFiles"
+getFiles :: List FilePath -> IO (List (FilePath, Chars))
+getFiles paths = execIo getFile paths
 
 -- Given a file name, read it and for each line in that file, read and print contents of each.
 -- Use @getFiles@ and @printFiles@.
-run ::
-  FilePath
-  -> IO ()
-run =
-  error "todo: Course.FileIO#run"
+run :: FilePath -> IO ()
+run indexFilePath =
+  readFile indexFilePath
+  >>= \chars -> getFiles (lines chars)
+  >>= printFiles
 
 -- /Tip:/ use @getArgs@ and @run@
-main ::
-  IO ()
+main :: IO ()
 main =
-  error "todo: Course.FileIO#main"
+  getArgs
+  >>= \args -> execIoNoResult run args
 
-----
 
 -- Was there was some repetition in our solution?
 -- ? `sequence . (<$>)`
 -- ? `void . sequence . (<$>)`
 -- Factor it out.
+
+-- Run a bunch of (t -> IO ()) on a List (t), reduce to IO () result
+execIoNoResult :: (t -> IO ()) -> List t -> IO ()
+execIoNoResult f xs = void $ execIo f xs
```

Comment:
This could be rewritten `execIONoResult = void . execIO`. Or, if you really want to do primitive recursion, at least use a fold... but it's better to use the abstractions we already have.

It feels like this function is kind of redundant. If we know what `sequence` does, and `void` does, then we don't need this helper function, right? it's just `void . sequence`. In "actual haskell" this function is already defined.

Also, it's *not* executing anything... seems like a bit of a minor nitpick, but it's actually really important. We're *composing a single action out of many actions*. Which is to say, we're producing a single effectful value out of many effectful values. It really **is** *sequencing the applicative effect*.

Reply:
It's hard to name functions when you're only vaguely aware of what's going on. I definitely don't pick very good names for things. 

The reason I wrote these functions at the bottom was because the comments in the course said:

```
-- Was there was some repetition in our solution?
-- ? `sequence . (<$>)`
-- ? `void . sequence . (<$>)`
-- Factor it out.
```

So I was trying to do that.

What would you call `void . sequence` if you had to pick a name?

Reply:
```
>> :t void . sequence . (<$>)

<interactive>:1:19: error:
    • Couldn't match type ‘f0 a -> f0 b’ with ‘List (f a0)’
      Expected type: (a -> b) -> List (f a0)
        Actual type: (a -> b) -> f0 a -> f0 b
    • Probable cause: ‘(<$>)’ is applied to too few arguments
      In the second argument of ‘(.)’, namely ‘(<$>)’
      In the second argument of ‘(.)’, namely ‘sequence . (<$>)’
      In the expression: void . sequence . (<$>)
```

Maybe not that one?

Reply:
The other suggestion also blows up with type errors. Not sure what they're asking.

Reply:
Ohhhh I see... I didn't really understand that it meant across functions. Interesting :)

Yeah sorry about giving you a hard time about naming, especially when you haven't yet quite got a grip on what it means. Like... why is `sequence` that? it's a bit of a weird name for something that sequences the applicative effects of a list of applicatives.

And especially when Haskell is so inconsistent with itself about naming.

I usually use fairly boring names for helper functions in Haskell:

```
sequenceToUnit :: Applicative f => List (f a) -> f ()
sequenceToUnit = void . sequence

```

I'm not sure what `something f = void . sequence . (<$>) f` is for (note you need to specify the function or you'll get that composition error you got)... given `void` throws away the content of the applicative context, and `<$>` maps  it to something else, it seems a bit pointless to map it then throw it away?

Reply:
Also, might be worth saying, it's good to be *less* specific about your types, generally, if you can... because it makes your functions more reusable, but also limits the possible definitions for them.

I'm not sure if this will be digestible yet, but it (possibly arrogantly) captures some more detail about why this is... http://degoes.net/articles/insufficiently-polymorphic

(so, rather than specify `IO a` maybe specify `Monad m => m a` or `Applicative f => f a`... whatever the most general interface you can get away with... )

Reply:
... then later, if you **really want** `sequenceIOToUnit` you can write it as:

```
sequenceIOToUnit :: List (IO a) -> IO ()
sequenceIOToUnit = sequenceToUnit
```

and it "just" becomes a type specialized re-definition of our `sequenceToUnit` function.